wrappers/internal/service_unit_gen.go

```diff
@@ -292,25 +301,36 @@ WantedBy={{.ServicesTarget}}
 		OOMAdjustScore: oomAdjustScore,
 		BusName:        busName,
 
-		Before: generateServiceNames(appInfo.Snap, appInfo.Before),
-		After:  generateServiceNames(appInfo.Snap, appInfo.After),
+		Before:  generateServiceNames(appInfo.Snap, appInfo.Before),
+		After:   generateServiceNames(appInfo.Snap, appInfo.After),
+		BindsTo: generateServiceNames(appInfo.Snap, appInfo.BindsTo),
+		PartOf:  generateServiceNames(appInfo.Snap, appInfo.PartOf),
 
 		// systemd runs as PID 1 so %h will not work.
 		Home: "/root",
 	}
-	switch appInfo.DaemonScope {
-	case snap.SystemDaemon:
+	switch {
+	case appInfo.DaemonScope.IsSystemDaemon():
 		wrapperData.ServicesTarget = systemd.ServicesTarget
 		wrapperData.PrerequisiteTarget = systemd.PrerequisiteTarget
 		wrapperData.MountUnit = filepath.Base(systemd.MountUnitPath(dirs.StripRootDir(appInfo.Snap.MountDir())))
 		wrapperData.Requires = append(wrapperData.Requires, wrapperData.MountUnit)
 		wrapperData.WorkingDir = dirs.StripRootDir(appInfo.Snap.DataDir())
 		wrapperData.After = append(wrapperData.After, "snapd.apparmor.service")
-	case snap.UserDaemon:
+	case appInfo.DaemonScope.IsUserDaemon():
 		wrapperData.ServicesTarget = systemd.UserServicesTarget
 		// FIXME: ideally use UserDataDir("%h"), but then the
 		// unit fails if the directory doesn't exist.
 		wrapperData.WorkingDir = appInfo.Snap.DataDir()
+		if appInfo.DaemonScope == snap.GraphicalUserDaemonScope {
+			wrapperData.After = append(wrapperData.After, "graphical-session.target")
+			if appInfo.Daemon == "dbus" || len(appInfo.Sockets) != 0 {
+				wrapperData.PartOf = append(wrapperData.PartOf, "graphical-session.target")
```

Comment:
Now... What if a daemon wants to be started up on login, but also be dbus activatable (many ui demons have idle timeouts, so they go down after a while...)

However I don't think this is a big deal in practice, just mentioning it in case there is some case we didn't think about. 😬

Reply:
Is that currently possible with the current snapd/snapcraft system...?